#191 Modified StateRepresentation.Exit to allow recursive Exit action…

HenningNT opened the pull request:
…s to be executed. It now modifies the Transition object as it executes Exit recursively. It returns a new transistion object when done, so that the source is set to the topmost superstate.

Modified StateMachine.InternalFireOne to accept the modified transition object returned by StateRepresentation.Exit.
Added needed test to verify correct handling of reeantrant transitions, as pointed out by @dpavlenkov

nblumhardt commented on test/Stateless.Tests/StateMachineFixture.cs:
```diff
@@ -442,5 +442,30 @@ public void IgnoreVsPermitReentryFrom()
 
             Assert.Equal(1, _numCalls);
         }
+        [Fact]
```
Nit: whitespace

nblumhardt commented on src/Stateless/StateRepresentation.cs:
```diff
@@ -176,7 +176,7 @@ public void Enter(Transition transition, params object[] entryArgs)
             }
 
-            public void Exit(Transition transition)
+            public Transition Exit(Transition transition)
```
Is there an `async` version of this method somewhere that we need to update?

HenningNT replied:
There's 'public async Task ExitAsync(Transition transition)'. I guess this has to modified to return Task<Transition>...

HenningNT commented:
I've modified ExitAsync and InternalFireOneAsync as well.
To be honest, I don't like the async part of StateMachine at all. To make the test pass I added await sm.FireAsync(Trigger.X); Otherwise the Entry and Exit actions were executed on different Tasks that didn't complete before the Assert were executed.

Adding await to sm.FireAsync forces the state machine to become synchronous, which sort of defeats the purpose of making an async version in the first place :-D
I really think that actions should be run atomically and run-to-completion.

dpavlenkov commented:
Indeed, my usage of any control logic (which is what SM is) is to wrap it. If it wants to do something async, it has be an action to start something and trigger to complete it. I love Stateless API and idea, but I'm using my own implementation with just bare essentials, thoroughly tested, zero async.

nblumhardt commented:
@HenningNT Looks good! 👍 

> Adding await to sm.FireAsync forces the state machine to become synchronous

The thread will still be freed up to do other work while awaiting the call, which is the main aim; the async version of the state machine doesn't add any concurrency.
